### PR TITLE
JAMES-3737 Reactive IMAP implementation

### PR DESCRIPTION
**Pull request by chibenwa**
Large throughtput gains

### Before

Screenshot from 2022-03-29 13-08-53

### After

Screenshot from 2022-03-29 13-09-42

### TODO

**MUST**

 - [ ] Better evalute latencies on low throughtput
 - [x] I would really like to reactify IMAP STORE command
 - [x] The following bench can further be improved by reactifying `SelectedMailboxImpl.deselect`. `jstack` thread analysis shows worker threads blocked on that operation.
 - [X] Preserve structured logging for IMAP

**COULD**

 - [ ] IMAP MOVE + IMAP COPY spaghetti code get's in the way too and would need a massive refactoring

### CODE REVIEW

**Comment by chibenwa** [comment]
Force pushed to solve a rebase conflict. I also added reactive code for mailbox deselection...

**Comment by chibenwa** [comment]
Screenshot from 2022-04-01 10-02-15

**Comment by chibenwa** [comment]
Arf I broke some JPA stuff :-/

```
org.apache.james.mpt.imapmailbox.jpa.JpaUidSearchTest.testSearchCombinationsITALY
org.apache.james.mpt.imapmailbox.jpa.JpaSearchTest.testSearchAtomsKOREA
org.apache.james.mpt.imapmailbox.jpa.JpaConcurrentSessionsTest.expungeShouldNotBreakUIDToMSNMapping
org.apache.james.mpt.imapmailbox.jpa.JpaConcurrentSessionsTest.testConcurrentExpungeResponseUS
org.apache.james.mpt.imapmailbox.jpa.JpaConcurrentSessionsTest.testConcurrentCrossExpungeKOREA
```

Todo: fix this!

File: protocols/imap/src/main/java/org/apache/james/imap/processor/ExpungeProcessor.java [hunk 1]
```diff
@@ -109,6 +84,36 @@ protected void processRequest(ExpungeRequest request, ImapSession session, Respo
         }
     }
 
+    private int expunge(ExpungeRequest request, ImapSession session, MessageManager mailbox, MailboxSession mailboxSession) throws MailboxException {
+        IdRange[] ranges = request.getUidSet();
+        if (ranges == null) {
+           return expunge(mailbox, MessageRange.all(), session, mailboxSession);
```
Comment (Arsnael) [inline protocols/imap/src/main/java/org/apache/james/imap/processor/ExpungeProcessor.java:94]: indent

**Review by Arsnael — COMMENTED** [review]
Small comments for now mainly :)

File: server/protocols/protocols-imap4/src/main/java/org/apache/james/imapserver/netty/ImapChannelUpstreamHandler.java [hunk 2]
```diff
@@ -196,7 +196,9 @@ public void channelInactive(ChannelHandlerContext ctx) throws Exception {
                 .orElse(Mono.empty())
                 .doFinally(signal -> imapConnectionsMetric.decrement())
                 .doFinally(Throwing.consumer(signal -> super.channelInactive(ctx)))
-                .subscribe(any -> {}, ctx::fireExceptionCaught);
+                .subscribe(any -> {
+
+                }, ctx::fireExceptionCaught);
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply server/protocols/protocols-imap4/src/main/java/org/apache/james/imapserver/netty/ImapChannelUpstreamHandler.java:202]: Yes, checkstyle...

File: mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/mail/task/MemoryRecomputeCurrentQuotasServiceTest.java [hunk 3]
```diff
@@ -1,87 +0,0 @@
...
-import org.apache.james.mailbox.inmemory.manager.InMemoryIntegrationResources;
-import org.apache.james.mailbox.quota.CurrentQuotaManager;
-import org.apache.james.mailbox.quota.UserQuotaRootResolver;
-import org.apache.james.mailbox.quota.task.RecomputeCurrentQuotasService;
-import org.apache.james.mailbox.quota.task.RecomputeCurrentQuotasServiceContract;
-import org.apache.james.user.api.UsersRepository;
-import org.apache.james.user.memory.MemoryUsersRepository;
-import org.junit.jupiter.api.BeforeEach;
-
-class MemoryRecomputeCurrentQuotasServiceTest implements RecomputeCurrentQuotasServiceContract {
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/mail/task/MemoryRecomputeCurrentQuotasServiceTest.java:38]: Mistake

**Review by Arsnael — COMMENTED** [review]
Nothing to add :)

**Comment by chibenwa** [comment]
Wooohoooo 🍏 

Will rebase this tomorrow!

Will also update benchmarks...

**Comment by chibenwa** [comment]
This work is becoming serious: I thus remove its draft status.

I will conduct a performance campaign on it to ensure the current status is good, in line with previous reports.

**Comment by chibenwa** [comment]
Screenshot from 2022-04-13 17-29-37

So far so good :muscle:

File: protocols/imap/src/main/java/org/apache/james/imap/processor/DeleteProcessor.java [hunk 4]
```diff
@@ -57,16 +57,19 @@ protected Mono<Void> processRequestReactive(DeleteRequest request, ImapSession s
             .then(unsolicitedResponses(session, responder, false))
             .then(Mono.fromRunnable(() -> okComplete(request, responder)))
             .then()
+            .doOnEach(logOnError(MailboxNotFoundException.class, e -> LOGGER.debug("Delete failed for mailbox {} as it doesn't exist", mailboxPath, e)))
             .onErrorResume(MailboxNotFoundException.class, e -> {
                 LOGGER.debug("Delete failed for mailbox {} as it doesn't exist", mailboxPath, e);
```
Comment (quantranhong1999) [inline protocols/imap/src/main/java/org/apache/james/imap/processor/DeleteProcessor.java:62]: We should remove this line? (already logged in doOnEach)

File: protocols/imap/src/main/java/org/apache/james/imap/processor/DeleteProcessor.java [hunk 5]
```diff
@@ -57,16 +57,19 @@ protected Mono<Void> processRequestReactive(DeleteRequest request, ImapSession s
...
                 no(request, responder, HumanReadableText.FAILURE_NO_SUCH_MAILBOX);
                 return Mono.empty();
             })
+            .doOnEach(logOnError(TooLongMailboxNameException.class, e -> LOGGER.debug("The mailbox name length is over limit: {}", mailboxPath.getName(), e)))
             .onErrorResume(TooLongMailboxNameException.class, e -> {
                 LOGGER.debug("The mailbox name length is over limit: {}", mailboxPath.getName(), e);
```
Comment (quantranhong1999) [inline protocols/imap/src/main/java/org/apache/james/imap/processor/DeleteProcessor.java:68]: idem

File: protocols/imap/src/main/java/org/apache/james/imap/processor/DeleteProcessor.java [hunk 6]
```diff
@@ -57,16 +57,19 @@ protected Mono<Void> processRequestReactive(DeleteRequest request, ImapSession s
...
                 taggedBad(request, responder, HumanReadableText.FAILURE_MAILBOX_NAME);
                 return Mono.empty();
             })
+            .doOnEach(logOnError(MailboxException.class, e -> LOGGER.error("Delete failed for mailbox {}", mailboxPath, e)))
             .onErrorResume(MailboxException.class, e -> {
                 LOGGER.error("Delete failed for mailbox {}", mailboxPath, e);
```
Comment (quantranhong1999) [inline protocols/imap/src/main/java/org/apache/james/imap/processor/DeleteProcessor.java:74]: idem

**Review by quantranhong1999 — COMMENTED** [review]
Great work to me!

**Review by Arsnael — APPROVED** [review]

**Review by Arsnael — APPROVED** [review]

**Comment by chibenwa** [comment]
:green_apple: . Objections?

**Comment by chibenwa** [comment]
Happy to fix remarks in follow up PRs.

**Pull request merged**